File: slp/backend-dotnet/Features/Question/QuestionService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;
using backend_dotnet.Features.Tag;
using backend_dotnet.Helpers;   // <-- ADDED

namespace backend_dotnet.Features.Question;

public class QuestionService : IQuestionService
{
    private readonly IQuestionRepository _questionRepository;
    private readonly ITagRepository _tagRepository;

    public QuestionService(IQuestionRepository questionRepository, ITagRepository tagRepository)
    {
        _questionRepository = questionRepository;
        _tagRepository = tagRepository;
    }

    public async Task<QuestionDto?> GetQuestionByIdAsync(int id)
    {
        var question = await _questionRepository.GetByIdAsync(id);
        return question == null ? null : MapToDto(question);
    }

    public async Task<IEnumerable<QuestionListDto>> GetUserQuestionsAsync(int userId)
    {
        var questions = await _questionRepository.GetUserQuestionsAsync(userId);
        return questions.Select(MapToListDto);
    }

    public async Task<IEnumerable<QuestionListDto>> GetAllQuestionsAsync()
    {
        var questions = await _questionRepository.GetAllQuestionsAsync();
        return questions.Select(MapToListDto);
    }

    public async Task<QuestionDto> CreateQuestionAsync(int userId, CreateQuestionDto dto)
    {
        // Validate metadata based on type
        QuestionValidationHelper.ValidateQuestionMetadata(dto.Type, dto.Content, dto.MetadataJson);   // <-- CHANGED

        var question = new Question
        {
            UserId = userId,
            Type = dto.Type,
            Content = dto.Content,
            Explanation = dto.Explanation,
            MetadataJson = dto.MetadataJson,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        if (dto.TagNames != null && dto.TagNames.Any())
        {
            var tags = await _tagRepository.GetOrCreateTagsAsync(dto.TagNames);
            question.QuestionTags = tags.Select(t => new QuestionTag { Tag = t }).ToList();
        }

        var created = await _questionRepository.CreateAsync(question);
        return MapToDto(created);
    }

    public async Task<QuestionDto?> UpdateQuestionAsync(int id, int userId, UpdateQuestionDto dto)
    {
        var question = await _questionRepository.GetByIdAsync(id);
        if (question == null || question.UserId != userId)
            return null;

        // Determine the effective type after update (if type is being changed)
        string effectiveType = dto.Type ?? question.Type;

        // If metadata is being updated, validate it using the effective type and content
        if (dto.MetadataJson != null)
        {
            // Content might also be updated; use the new content if provided
            string effectiveContent = dto.Content ?? question.Content;
            QuestionValidationHelper.ValidateQuestionMetadata(effectiveType, effectiveContent, dto.MetadataJson);   // <-- CHANGED
        }

        question.Type = dto.Type ?? question.Type;
        question.Content = dto.Content ?? question.Content;
        question.Explanation = dto.Explanation ?? question.Explanation;
        question.MetadataJson = dto.MetadataJson ?? question.MetadataJson;

        if (dto.TagNames != null)
        {
            _tagRepository.RemoveQuestionTags(question.Id);
            var tags = await _tagRepository.GetOrCreateTagsAsync(dto.TagNames);
            question.QuestionTags = tags.Select(t => new QuestionTag { QuestionId = question.Id, TagId = t.Id }).ToList();
        }

        await _questionRepository.UpdateAsync(question);
        return MapToDto(question);
    }

    public async Task<bool> DeleteQuestionAsync(int id, int userId, bool isAdmin)
    {
        var question = await _questionRepository.GetByIdAsync(id);
        if (question == null)
            return false;

        if (!isAdmin && question.UserId != userId)
            return false;

        await _questionRepository.SoftDeleteAsync(id);
        return true;
    }

    public async Task<IEnumerable<QuestionListDto>> SearchQuestionsAsync(QuestionSearchDto search)
    {
        var questions = await _questionRepository.SearchAsync(search.SearchTerm, search.Type, search.Tags, search.UserId);
        return questions.Select(MapToListDto);
    }

    // ------------------------------------------------------------------------
    // Mapping helpers (unchanged)
    // ------------------------------------------------------------------------
    private QuestionDto MapToDto(Question q)
    {
        return new QuestionDto
        {
            Id = q.Id,
            UserId = q.UserId,
            Type = q.Type,
            Content = q.Content,
            Explanation = q.Explanation,
            MetadataJson = q.MetadataJson,
            CreatedAt = q.CreatedAt,
            UpdatedAt = q.UpdatedAt,
            Tags = q.QuestionTags?.Select(qt => qt.Tag?.Name ?? "").ToList() ?? new List<string>(),
            UserName = q.User?.Username
        };
    }

    private QuestionListDto MapToListDto(Question q)
    {
        return new QuestionListDto
        {
            Id = q.Id,
            Type = q.Type,
            Content = q.Content,
            Explanation = q.Explanation,
            CreatedAt = q.CreatedAt,
            UpdatedAt = q.UpdatedAt,
            Tags = q.QuestionTags?.Select(qt => qt.Tag?.Name ?? "").ToList() ?? new List<string>(),
            UserName = q.User?.Username
        };
    }
}

File: slp/backend-nestjs/src/modules/question/question.service.ts
import { Injectable, NotFoundException, ForbiddenException, BadRequestException } from '@nestjs/common';
import { QuestionRepository } from './question.repository';
import { TagService } from '../tag/tag.service';
import { Question } from './question.entity';
import { QuestionTag } from './question-tag.entity';
import { QuestionDto, QuestionListDto } from './dto/question.dto';
import { CreateQuestionDto } from './dto/create-question.dto';
import { UpdateQuestionDto } from './dto/update-question.dto';
import { QuestionSearchDto } from './dto/question-search.dto';
import { PaginatedResult, paginate } from '../../helpers/pagination.helper';
import { QuestionValidationHelper } from './question-validation.helper';

@Injectable()
export class QuestionService {
  constructor(
    private questionRepo: QuestionRepository,
    private tagService: TagService,
  ) {}

  async getQuestionById(id: number): Promise<QuestionDto | null> {
    const question = await this.questionRepo.findById(id);
    return question ? this.mapToDto(question) : null;
  }

  async getUserQuestions(userId: number): Promise<QuestionListDto[]> {
    const questions = await this.questionRepo.findByUserId(userId);
    return questions.map(q => this.mapToListDto(q));
  }

  async getAllQuestions(): Promise<QuestionListDto[]> {
    const questions = await this.questionRepo.findAll();
    return questions.map(q => this.mapToListDto(q));
  }

  async createQuestion(userId: number, dto: CreateQuestionDto): Promise<QuestionDto> {
    // Validate metadata
    QuestionValidationHelper.validateQuestionMetadata(dto.type, dto.content, dto.metadataJson);

    const question = new Question();
    question.userId = userId;
    question.type = dto.type;
    question.content = dto.content;
    question.explanation = dto.explanation ?? null;
    question.metadataJson = dto.metadataJson ?? null;
    question.createdAt = new Date();
    question.updatedAt = new Date();

    if (dto.tagNames && dto.tagNames.length > 0) {
      const tags = await this.tagService.getOrCreateTags(dto.tagNames);
      question.questionTags = tags.map(tag => {
        const qt = new QuestionTag();
        qt.tagId = tag.id;
        return qt;
      });
    } else {
      question.questionTags = [];
    }

    const created = await this.questionRepo.createQuestion(question);
    return this.mapToDto(created);
  }

  async updateQuestion(id: number, userId: number, dto: UpdateQuestionDto): Promise<QuestionDto | null> {
    const question = await this.questionRepo.findById(id);
    if (!question) return null;
    if (question.userId !== userId) throw new ForbiddenException('You are not the owner of this question');

    // Determine effective type and content for validation
    const effectiveType = dto.type ?? question.type;
    const effectiveContent = dto.content ?? question.content;

    if (dto.metadataJson !== undefined) {
      QuestionValidationHelper.validateQuestionMetadata(effectiveType, effectiveContent, dto.metadataJson);
    }

    if (dto.type) question.type = dto.type;
    if (dto.content) question.content = dto.content;
    if (dto.explanation !== undefined) question.explanation = dto.explanation;
    if (dto.metadataJson !== undefined) question.metadataJson = dto.metadataJson;

    if (dto.tagNames) {
      await this.questionRepo.removeQuestionTags(question.id);
      const tags = await this.tagService.getOrCreateTags(dto.tagNames);
      question.questionTags = tags.map(tag => {
        const qt = new QuestionTag();
        qt.tagId = tag.id;
        return qt;
      });
    }

    const updated = await this.questionRepo.updateQuestion(question);
    return this.mapToDto(updated);
  }

  async deleteQuestion(id: number, userId: number, isAdminUser: boolean): Promise<boolean> {
    const question = await this.questionRepo.findById(id);
    if (!question) return false;
    if (!isAdminUser && question.userId !== userId) return false;
    await this.questionRepo.deleteQuestion(id);
    return true;
  }

  async searchQuestions(search: QuestionSearchDto, page: number = 1, pageSize: number = 20): Promise<PaginatedResult<QuestionListDto>> {
    const { items, totalCount } = await this.questionRepo.search(
      search.searchTerm,
      search.type,
      search.tags,
      search.userId,
      page,
      pageSize,
    );
    const dtos = items.map(q => this.mapToListDto(q));
    return paginate(dtos, totalCount, page, pageSize);
  }

  // Mapping helpers
  private mapToDto(q: Question): QuestionDto {
    return {
      id: q.id,
      userId: q.userId,
      type: q.type,
      content: q.content,
      explanation: q.explanation,
      metadataJson: q.metadataJson,
      createdAt: q.createdAt,
      updatedAt: q.updatedAt,
      tags: q.questionTags?.map(qt => qt.tag?.name).filter(Boolean) ?? [],
      userName: q.user?.username,
    };
  }

  private mapToListDto(q: Question): QuestionListDto {
    return {
      id: q.id,
      type: q.type,
      content: q.content,
      explanation: q.explanation,
      createdAt: q.createdAt,
      updatedAt: q.updatedAt,
      tags: q.questionTags?.map(qt => qt.tag?.name).filter(Boolean) ?? [],
      userName: q.user?.username,
      metadataJson: q.metadataJson,
    };
  }
}